### PR TITLE
Update dtls max version for dtls1.3

### PR DESCRIPTION
**Pull request by fwh-dc**
Another dtls 1.3 specific update. It is dependent on https://github.com/openssl/openssl/pull/22259

### CODE REVIEW

File: ssl/statem/statem_clnt.c [hunk 1]
```diff
@@ -191,7 +191,7 @@ static int ossl_statem_client13_read_transition(SSL_CONNECTION *s, int mt)
         if (mt == SSL3_MT_CERTIFICATE_REQUEST) {
 #if DTLS_MAX_VERSION_INTERNAL != DTLS1_2_VERSION
             /* Restore digest for PHA before adding message.*/
-# error Internal DTLS version error
+# warning Internal DTLS version error
 #endif
```
Comment (mattcaswell) [inline ssl/statem/statem_clnt.c:195]: Shouldn't this if/endif just be deleted?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: ssl/statem/statem_clnt.c (hunk 1 above)
Comment (mattcaswell) [reply ssl/statem/statem_clnt.c:195]: I actually think this TODO is now obsolete.

It was originally added in 9d75dce3e. At that time the read state machine code looked like this:

https://github.com/openssl/openssl/blob/9d75dce3e1f94be6006500089491cb3284f70d06/ssl/statem/statem.c#L565-L624

Note in this snippet that near the top, for TLS, we get the message header. In the middle we transition the state machine state and then near the end we read the TLS message body. However for DTLS we just get the whole message near the top, and then transition after.

Importantly in `tls_get_message_body()` we call `ssl3_finish_mac` to feed the message into the handshake digest - and this happens *after* the state machine has transitioned. However is DTLS it all happens in `dtls_get_message` which happens before the transition. This is what the comment in this function a little below this is talking about.

Fast forward to today and the code looks different. This is the equivalent code in the read state machine:

https://github.com/openssl/openssl/blob/40a24c20a809916b43116c2bb16a36bdc40221f3/ssl/statem/statem.c#L616-L673

In this code we still get the DTLS message all in one go at the top, but there is now a new function after the transition called `dtls_get_message_body`...it doesn't actually get the body but it does do post-processing of the body including feeding the message into the handshake hash, so that the message is hashed *after* the transition in the same way as for TLS.

In other words DTLS and TLS now work the same way in this regards and therefore we don't need to do anything special for DTLSv1.3 here.

Comment (fwh-dc) [reply ssl/statem/statem_clnt.c:195]: So if I understand you correctly this
`
#if DTLS_MAX_VERSION_INTERNAL != DTLS1_2_VERSION
            /* Restore digest for PHA before adding message.*/
# warning Internal DTLS version error
#endif
            if (!SSL_CONNECTION_IS_DTLS(s)
                && s->post_handshake_auth == SSL_PHA_EXT_SENT) {
                s->post_handshake_auth = SSL_PHA_REQUESTED;
                /*
                 * In TLS, this is called before the message is added to the
                 * digest. In DTLS, this is expected to be called after adding
                 * to the digest. Either move the digest restore, or add the
                 * message here after the swap, or do it after the clientFinished?
                 */
                if (!tls13_restore_handshake_digest_for_pha(s)) {
                    /* SSLfatal() already called */
                    return 0;
                }
                st->hand_state = TLS_ST_CR_CERT_REQ;
                return 1;
            }
`
Should be:
`
            if (s->post_handshake_auth == SSL_PHA_EXT_SENT) {
                s->post_handshake_auth = SSL_PHA_REQUESTED;
                if (!tls13_restore_handshake_digest_for_pha(s)) {
                    /* SSLfatal() already called */
                    return 0;
                }
                st->hand_state = TLS_ST_CR_CERT_REQ;
                return 1;
            }
`
?

Comment (mattcaswell) [reply ssl/statem/statem_clnt.c:195]: Yes, I believe so.

**Review by mattcaswell — APPROVED** [review]

**Comment by fwh-dc** [comment]
Test failures are relevant. We need to merge a couple of other PR's before this is working.

**Comment by mattcaswell** [comment]
> Test failures are relevant. We need to merge a couple of other PR's before this is working.

Will those failure go away with a rebase now?

**Comment by fwh-dc** [comment]
> > Test failures are relevant. We need to merge a couple of other PR's before this is working.
> 
> Will those failure go away with a rebase now?

No. We need https://github.com/openssl/openssl/pull/22364 to be merged first. Which is dependent on https://github.com/openssl/openssl/pull/22366.

And then I expect there are some test cases that needs to be limited to DTLS 1.2 for it to work. I'll add that work to this PR to fix the pipeline.

**Comment by fwh-dc** [comment]
@t8m @mattcaswell 

A quick update on this.

Please prioritise reviewing the following:
https://github.com/openssl/openssl/pull/22936
https://github.com/openssl/openssl/pull/22401
https://github.com/openssl/openssl/pull/22380 
https://github.com/openssl/openssl/pull/22360 

When they are merged I'll do a rebase of this branch and make some adjustments to make the tests pass (some has to be forced to run on DTLS 1.2 for now). That will enable the dtls 1.3 feature branch to negotiate dtls 1.3. And then we can make the required adjustments to make it correct according to the RFC.

**Comment by mattcaswell** [comment]
> Please prioritise reviewing the following:

All of these are now merged

**Comment by fwh-dc** [comment]
Yes thanks. I'll update this PR today or tomorrow.

**Comment by mattcaswell** [comment]
Note that #24161 is going to cause problems for this one. My current plan is:

Rebase the feature/dtls-1.3 branch later today on master (there will be no conflicts from this)
Merge #24161 to master
Once #24226 is approved I can rebase the feature/dtls-1.3 branch again which will bring #24161 onto this feature branch

At that point you will need to update this PR again to fix any resulting issues.

**Comment by fwh-dc** [comment]
@mattcaswell Please review again. I've marked a couple of tests for reenablement and some has been forced to DTLS 1.2. Please let me know what you think.

**Comment by t8m** [comment]
At least some of the CI failures look relevant.

File: ssl/statem/extensions_clnt.c [hunk 2]
```diff
@@ -17,12 +17,10 @@ EXT_RETURN tls_construct_ctos_renegotiate(SSL_CONNECTION *s, WPACKET *pkt,
                                           size_t chainidx)
 {
     if (!s->renegotiate) {
+        /**
+         * TODO(DTLSv1.3): Fix this when resumption and renegotiation works
+         */
```
Comment (mattcaswell) [inline ssl/statem/extensions_clnt.c:22]: Isn't it just this?:

```suggestion
        if (SSL_CONNECTION_IS_DTLS(s)
                && ssl_version_cmp(s->min_proto_version, DTLS1_3_VERSION) >= 0)
            return EXT_RETURN_NOT_SENT;
```

The renegotiation extension is not relevant for DTLSv1.3 since it doesn't support renegotiation.

File: ssl/statem/statem_lib.c [hunk 3]
```diff
@@ -2190,7 +2193,7 @@ int ssl_choose_server_version(SSL_CONNECTION *s, CLIENTHELLO_MSG *hello,
 
     if (suppversions->present) {
         unsigned int candidate_vers = 0;
-        const unsigned int best_vers_init = SSL_CONNECTION_IS_DTLS(s) ? UINT_MAX
+        const unsigned int best_vers_init = SSL_CONNECTION_IS_DTLS(s) ? INT_MAX
```
Comment (mattcaswell) [inline ssl/statem/statem_lib.c:2196]: Why this change?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: ssl/statem/extensions_clnt.c (hunk 2 above)
Comment (fwh-dc) [reply ssl/statem/extensions_clnt.c:22]: I thought so too. But then I get failures like:
```
    # 0041575DF87F0000:error:0A000152:SSL routines:final_renegotiate:unsafe legacy renegotiation disabled:../ssl/statem/extensions.c:946:
```

I haven't spent time looking at what goes wrong yet.

File: ssl/statem/statem_lib.c (hunk 3 above)
Comment (fwh-dc) [reply ssl/statem/statem_lib.c:2196]: Because version numbers are treated as int's in `ssl_version_cmp`. So actually we should change the type here?

Comment (mattcaswell) [reply ssl/statem/statem_lib.c:2196]: Yeah - that would be ok and avoids the mismatch between using the signed max and the unsigned type.

File: ssl/statem/extensions_clnt.c (hunk 2 above)
Comment (mattcaswell) [reply ssl/statem/extensions_clnt.c:22]: Yes that seems reasonable - although shouldn't it be:

```` C
    if (!ossl_assert(versiona > 0) || !ossl_assert(versionb > 0))
        return 0;
````

Comment (fwh-dc) [reply ssl/statem/extensions_clnt.c]: Good suggestion!

I think I want to do it as a seperate PR since we will need to fix checks in other places. Is there a way to get a full stack dump when ossl_assert is hit?

Comment (mattcaswell) [reply ssl/statem/extensions_clnt.c]: Not as far as I know. I usually just put it in the debugger.

**Review by mattcaswell — APPROVED** [review]

**Comment by mattcaswell** [comment]
CI failures are relevant

File: ssl/statem/extensions_clnt.c (hunk 2 above)
Comment (fwh-dc) [reply ssl/statem/extensions_clnt.c]: I've fixed the behaviour of ssl_version_cmp in
https://github.com/openssl/openssl/pull/24293 (feature/dtls-1.3)
https://github.com/openssl/openssl/pull/24292 (master)

Please have a look. I'd like to rebase this branch on those changes.

**Comment by mattcaswell** [comment]
Still some relevant CI failures by the look of it

**Comment by fwh-dc** [comment]
> Still some relevant CI failures by the look of it

Agreed, but less than last commit :)

I'll let you know when I think it is ready for review. Sorry for the wait.

**Comment by fwh-dc** [comment]
@mattcaswell @t8m 

I think I found a fix for the memory leaks, but the patch in [6adbb30](https://github.com/openssl/openssl/pull/22275/commits/6adbb3057df6b3caaf67fc1ec1ed660bcc5225db) seems to break compilation. TBH I don't understand why. Maybe you know what's causing it?

**Comment by mattcaswell** [comment]
It is because the test file test/tls13secretstest.c mocks outs some internal functions and links in ssl/tls13_enc.c directly:

https://github.com/openssl/openssl/blob/deaa83af700113c99835a1db7d45d33baba05bd3/test/build.info#L1028-L1038

To fix it you will need to supply dummy implementations of `dtls1_clear_received_buffer` and `dtls1_clear_sent_buffer` in tls13secretstest.c:

```` c
void dtls1_clear_received_buffer(SSL_CONNECTION *s)
{
}

void dtls1_clear_sent_buffer(SSL_CONNECTION *s)
{
}
````

**Comment by fwh-dc** [comment]
@mattcaswell @t8m 

This is ready for review once again. I have added some handling of middlebox compatibility and fixed the memory leak. Please let me know what you think.

**Review by mattcaswell — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by mattcaswell** [comment]
Pushed. Thanks.
